nats/rtt_command.go

```diff
@@ -0,0 +1,105 @@
+// Copyright 2020 The NATS Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"crypto/tls"
+	"fmt"
+	"log"
+	"net"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/nats-io/nats.go"
+	"gopkg.in/alecthomas/kingpin.v2"
+)
+
+func configureRTTCommand(app *kingpin.Application) {
+	app.Command("rtt", "Compute round-trip time to NATS server").Action(rtt)
+}
+
+func rtt(pc *kingpin.ParseContext) error {
+	if !strings.Contains(servers, "://") {
+		servers = fmt.Sprintf("nats://%s", servers)
+	}
+
+	u, err := url.Parse(servers)
+	if err != nil {
+		return err
+	}
+
+	if net.ParseIP(u.Hostname()) == nil {
+		addrs, _ := net.LookupHost(u.Hostname())
+		if len(addrs) <= 1 {
+			goto doSingle
+		}
+		opts := natsOpts()
+		nc, err := newNatsConn(servers, opts...)
+		if err != nil {
+			return err
+		}
+		log.Printf("[%s]\n", nc.ConnectedUrl())
+		if nc.TLSRequired() {
+			opts = append(opts, useTLSName(u.Hostname()))
+		}
+		nc.Close()
+		for _, a := range addrs {
+			server := fmt.Sprintf("%s://%s", u.Scheme, net.JoinHostPort(a, u.Port()))
+			_, rtt, err := calcRTT(server, opts)
+			if err != nil {
+				return err
+			}
+			log.Printf("    [%-15s] = avg %v\n", a, rtt)
+		}
+		return nil
+	}
+
+doSingle:
+	url, rtt, err := calcRTT(servers, natsOpts())
+	if err != nil {
+		return err
+	}
+	log.Printf("[%s] = avg %v\n", url, rtt)
+
+	return nil
+}
+
+func useTLSName(name string) nats.Option {
```

Comment:
can we put these methods - all except the configureRTTCommand on a struct? there's like 120 functions in this package so I am trying to keep related ones together, helps a bit with browsing code in IDE etc

Reply:
Sure.